### PR TITLE
ADFineGrainedPasswordPolicy: New resource

### LINKED ISSUE #584: ADFineGrainedPasswordPolicy: New resource proposal
### Description
New resource to support fine grained password policies for AD Principals.

### Proposed properties
The new resource will be using the Active Directory module commands below:

```powershell
Get-ADFineGrainedPasswordPolicy
Get-ADFineGrainedPasswordPolicySubject
New-ADFineGrainedPasswordPolicy
Set-ADFineGrainedPasswordPolicy
Remove-ADFineGrainedPasswordPolicy
Add-ADFineGrainedPasswordPolicySubject
Remove-ADFineGrainedPasswordPolicySubject
```

The resource will contain the following properties:

| Parameter | Attribute | DataType | Description | Allowed Values |
| --- | --- | --- | --- | --- |
| Name | Key | String | Specifies an Active Directory fine-grained password policy object name.
| Precedence | Required | UInt32 | Specifies a value that defines the precedence of a fine-grained password policy among all fine-grained password policies.
| DisplayName | Write | String | Specifies the display name of the object.
| Description | Write | String | Specifies a description for the object.
| Subjects | Write | String Array | Specifies the ADPrincipal names the policy is to be applied to, overwrites all existing.
| Ensure | Write | String | Specifies whether the fine-grained password policy should be present or absent. Default value is 'Present'. | "Present", "Absent" |
| ComplexityEnabled | Write | Boolean | Specifies whether password complexity is enabled for the password policy.
| LockoutDuration | Write | String | Specifies the length of time that an account is locked after the number of failed login attempts exceeds the lockout threshold. The lockout duration must be greater than or equal to the lockout observation time for a password policy. The value must be a string representation of a TimeSpan value.
| LockoutObservationWindow | Write | String | Specifies the maximum time interval between two unsuccessful login attempts before the number of unsuccessful login attempts is reset to 0. The lockout observation window must be smaller than or equal to the lockout duration for a password policy. The value must be a string representation of a TimeSpan value.
| LockoutThreshold | Write | UInt32 | Specifies the number of unsuccessful login attempts that are permitted before an account is locked out.
| MinPasswordAge | Write | String | Specifies the minimum length of time before you can change a password. The value must be a string representation of a TimeSpan value.
| MaxPasswordAge | Write | String | Specifies the maximum length of time that you can have the same password. The value must be a string representation of a TimeSpan value.
| MinPasswordLength | Write | UInt32 | Specifies the minimum number of characters that a password must contain.
| PasswordHistoryCount | Write | UInt32 | Specifies the number of previous passwords to save.
| ReversibleEncryptionEnabled | Write |Boolean | Specifies whether the directory must store passwords using reversible encryption.
| ProtectedFromAccidentalDeletion | Write | Boolean | Specifies whether to prevent the object from being deleted.
| DomainController | Write | String | Specifies the Active Directory Domain Services instance to connect to.
| Credential | Write | MSFT_Credential | Specifies the user account credentials to use to perform this task.

### PR DESCRIPTION
**Pull request by fl0p3k**
#### Pull Request (PR) description

This PR adds the `ADFineGrainedPasswordPolicy` resource to the module.

#### This Pull Request (PR) fixes the following issues

- Fixes #584

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [x] Resource documentation added/updated in README.md.
- [x] Resource parameter descriptions added/updated in README.md, schema.mof
      and comment-based help.
- [x] Comment-based help added/updated.
- [x] Localization strings added/updated in all localization files as appropriate.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

---
This change is [Reviewable](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585)

### CODE REVIEW

**Comment by johlju** [comment]
@fl0p3k can you please look at the failing tests?

**Comment by johlju** [comment]
Please see @X-Guardian's https://github.com/dsccommunity/ActiveDirectoryDsc/issues/584#issuecomment-610797852

**Comment by fl0p3k** [comment]
I am going to push the unit tests today once completed coding for it

**Comment by X-Guardian** [comment]
Ian, please can you complete the new resource proposal issue as requested before continuing with this PR, so that we can discuss the properties and implementation of this resource.

**Comment by fl0p3k** [comment]
> Ian, please can you complete the new resource proposal issue as requested before continuing with this PR, so that we can discuss the properties and implementation of this resource.

I have just updated the proposal #584

**Comment by fl0p3k** [comment]
> @fl0p3k can you please look at the failing tests?

I corrected the failing tests

**Comment by fl0p3k** [comment]
I will be looking over the comments and will provide some updates.

**Comment by fl0p3k** [comment]
*[a discussion](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-M4bcKXZ9E1jzLi9GSiU:-M57iJmc1FzXawIUMQa7:bifattt) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please resolve the comments in the issue, or discuss them if you have other thoughts.
</blockquote></details>

I have made all changes from the review feedback.  The only remaining pending action would be to understand the changes needed to support the PsDscRunAsCredential modifications.  Right now all is working and not sure I need to change anything regarding that.

---

**Comment by X-Guardian** [comment]
Ok, as I said in the resource proposal issue, the `ADDomainDefaultPasswordPolicy` resource was written four years ago, and does not use a modern logic pattern. It also not a resource that uses an 'Ensure' property, so the logic flow through the functions is very different.
A better example to follow is the `ADManagedServiceAccount` which has a modern logic flow of an 'Ensure' property resource both in the module and in its unit tests.

With regard to @johlju comments about the `Credential` parameter, I'm happy to leave this in, as it along with the `DomainController` parameter are needed to execute the resource on a non-domain joined machine, where the PsDscRunAsCredential parameter with domain credentials can't be used.

Key features that I would like the new resource to use:
- `Get-TargetResource` function only includes mandatory parameters and parameters explicitly required for the function (DomainController and Credential)
- `Get-TargetResource` function only returns properties defined in the schema.
- No need to use the imutablePropertyMap
- Try/Catch blocks with managed exceptions around every AD Cmdlet call.
- Use `Compare-ResourcePropertyState` common function in the `Test` and `Set` functions.
- Use the `Assert-MemberParameters` common function in the `Test` and `Set` functions to verify the  parameter combination given is valid.
- Examples in the relevant source/Examples/Resources folder with valid script headers.
- Property descriptions as specified in the resource proposal.
- Keep all line lengths to a maximum of 120 characters
- 100% unit test code coverage.

If you are not sure about anything, please ask before coding!

**Comment by fl0p3k** [comment]
@X-Guardian 
I've been away from this project for a bit, but I have implemented some changes based on your feedback.  I am hoping that I am meeting the bar at this point to move forward.  My only doubt was how to determine the code coverage the unit tests were evaluating.  I did my best to ensure that all scenarios would get tested.  Thanks.

**Comment by X-Guardian** [comment]
Hi @fl0p3k, If you go to the 'Checks' tab on this PR and then select the top check (`dsccommunity.ActiveDirectoryDsc`) then "View more details on Azure Pipelines" then select "Code Coverage", you will see the Code Coverage report for the module. If you then drill down into the `MSFT_ADFineGrainedPasswordPolicy` module you will get a line-by-line report of coverage.

Taking a quick look at your updated code, we are getting closer, but there are still some of the features that I listed that haven't been implemented:

- `Get-TargetResource` function only includes mandatory parameters and parameters explicitly required for the function (DomainController and Credential).
- `Get-TargetResource` function only returns properties defined in the schema.
- Use `Compare-ResourcePropertyState` common function in the Test and Set functions. Currently only used in the Test function.
- Use the `Assert-MemberParameters` common function in the Test and Set functions to verify the parameter combination given is valid.
- Examples in the relevant source/Examples/Resources folder with valid script headers.
- Keep all line lengths to a maximum of 120 characters.
- 100% unit test code coverage.

**Comment by fl0p3k** [comment]
> * `Get-TargetResource` function only includes mandatory parameters and parameters explicitly required for the function (DomainController and Credential).
> * `Get-TargetResource` function only returns properties defined in the schema.

@X-Guardian 
Ok, the two items above I am not sure I am on the same page.  I updated the Get-TargetResource function parameters that no longer use the DomainController or Credential parameters.  Also, I updated the Schema to include the SubjectsDifferent read only boolean output.

**Comment by X-Guardian** [comment]
Hi @fl0p3k , `Get-TargetResource` needs the `DomainController` and `Credential` parameters to pass to the `Get-ADFineGrainedPasswordPolicy` cmdlet, but it doesn't need the `Subjects`, `SubjectsToInclude` and `SubjectsToExclude` parameters. There is also no need for the `SubjectsDifferent` read only property, which we didn't agree to in the resource proposal. Please stick to that, or discuss first if you think you need to change it.

**Comment by fl0p3k** [comment]
> Hi @fl0p3k , `Get-TargetResource` needs the `DomainController` and `Credential` parameters to pass to the `Get-ADFineGrainedPasswordPolicy` cmdlet, but it doesn't need the `Subjects`, `SubjectsToInclude` and `SubjectsToExclude` parameters. There is also no need for the `SubjectsDifferent` read only property, which we didn't agree to in the resource proposal. Please stick to the proposal, or discuss first if you think you need to change it.

@X-Guardian, is there a forum we can chat in realtime (best time as well for you, I'm in PST)? I've been limited being able to focus on this but I have a few days ahead to address updates, but I would like to have a discussion about specific implements before making any more changes.

**Comment by X-Guardian** [comment]
Hi @fl0p3k , I'm available on the Slack PowerShell channel at http://slack.poshcode.org/

**Comment by fl0p3k** [comment]
Hmm, I think I need an invitation, I am getting a message that the link is no longer active.

**Comment by X-Guardian** [comment]
Try the link on the home page: https://poshcode.org/

**Comment by johlju** [comment]
@fl0p3k here is more information and links (let me know if they don't work ) https://dsccommunity.org/community/contact/

**Comment by fl0p3k** [comment]
I believe it is now ready with 100% code coverage and the changes to support the subjects.  Please review at your convenience.  Thanks!

**Comment by fl0p3k** [comment]
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 64 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXY9NL16w3VCzZllRe:-MAYOf5S2dmZFiGL-MM4:bso8c24) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L64)):*
<details><summary><i>Previously, X-Guardian (Simon Heather) wrote…</i></summary><blockquote>

Please remove the `Precedence` parameter from $parameters here, then there is no need to remove it from both `$getADFineGrainedPasswordPolicyParams` and `$getADFineGrainedPasswordPolicySubjectParams` further down.
</blockquote></details>

I get an error if I remove it since it is called out as mandatory in the schema.  This is a required parameter for the 'New-ADFineGrainedPasswordPolicy' cmdlet and I must make it mandatory in the schema I think.

**Comment by fl0p3k** [comment]
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 92 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXZEeQ7pd2Ixdd_uPE:-MAYPu8cEKJrO_9MEcpR:b-xj0d5m) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L92)):*
<details><summary><i>Previously, X-Guardian (Simon Heather) wrote…</i></summary><blockquote>

I only get an `ADIdentityNotFoundException` when the password policy doesn't exist, not a lack of subjects. I don't think you need this specific catch as it would have been caught in the previous catch.
</blockquote></details>

I added this incase something else goes wrong (like auth or server error from the parameters).

**Comment by fl0p3k** [comment]
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 318 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXaDug7aQaGejO1MB4:-MAYR01T-q6NojkFxr1S:bzdujxp) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L318)):*
<details><summary><i>Previously, X-Guardian (Simon Heather) wrote…</i></summary><blockquote>

What is this for?
</blockquote></details>

I mentioned this in the chat, but when the subjects are null from the Get-TargetResource function, the Compare-ResoucePropertyState does not detect anything mismatched.

**Comment by fl0p3k** [comment]
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 318 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXaDug7aQaGejO1MB4:-MAjQsfw1s8YWq2y2hsl:b-ufa0xj) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L318)):*
<details><summary><i>Previously, fl0p3k (Ian Florek) wrote…</i></summary><blockquote>

I have always returned an empty array when the subjects are null.  I made some specific changes to ensure only an empty array object is returned, but I get this:

VERBOSE: [IFLABDC01]:                            [[ADFineGrainedPasswordPolicy]UserAccounts] Evaluating the state of the property 'Subjects'. (ADCOMMON0003)
Cannot bind argument to parameter 'ReferenceObject' because it is null.
    + CategoryInfo          : InvalidData: (:) [], CimException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.CompareObjectCommand
    + PSComputerName        : IFLABDC01

VERBOSE: [IFLABDC01]:                            [[ADFineGrainedPasswordPolicy]UserAccounts] The parameter 'Subjects' is in desired state. (ADCOMMON0004)
</blockquote></details>

I figured this one out by using [string[]] instead of object array.

**Comment by fl0p3k** [comment]
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 318 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXaDug7aQaGejO1MB4:-MAjR2o_0NfJ73QvzJGs:b-896fix) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L318)):*
<details><summary><i>Previously, fl0p3k (Ian Florek) wrote…</i></summary><blockquote>

I figured this one out by using [string[]] instead of object array.
</blockquote></details>

Done.

**Comment by fl0p3k** [comment]
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 92 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXZEeQ7pd2Ixdd_uPE:-MB5MT7d5-IlkcK9holg:b-896fix) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L92)):*
<details><summary><i>Previously, X-Guardian (Simon Heather) wrote…</i></summary><blockquote>

Yes that would be better
</blockquote></details>

Done.

**Comment by fl0p3k** [comment]
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 92 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXZEeQ7pd2Ixdd_uPE:-MBeuONP5uwZvKFEJSkc:btqp1dl) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L92)):*
<details><summary><i>Previously, fl0p3k (Ian Florek) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Any update on closing this out and merging?

**Comment by johlju** [comment]
@x-guardian have probably been busy with the day job, I think he come around to this as soon as possible.

**Comment by X-Guardian** [comment]
Hi @johlju, I've been talking to @fl0p3k about this PR on Slack.

File: source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof [hunk 1]
```diff
@@ -0,0 +1,22 @@
+[ClassVersion("1.0.0"), FriendlyName("ADFineGrainedPasswordPolicy")]
+class MSFT_ADFineGrainedPasswordPolicy : OMI_BaseResource
+{
+    [Key, Description("Specifies an Active Directory fine-grained password policy object name.")] String Name;
+    [Required, Description("Specifies a value that defines the precedence of a fine-grained password policy among all fine-grained password policies.")] UInt32 Precedence;
+    [Write, Description("Specifies the display name of the object.")] String DisplayName;
+    [Write, Description("Specifies the ADPrincipal names the policy is to be applied to, overwrites all existing.")] String Subjects[];
+    [Write, Description("Specifies whether the fine-grained password policy should be present or absent. Default value is 'Present'."), ValueMap{"Present", "Absent"}, Values{"Present", "Absent"}] String Ensure;
+    [Write, Description("Specifies whether password complexity is enabled for the password policy.")] Boolean ComplexityEnabled;
+    [Write, Description("Specifies the length of time that an account is locked after the number of failed login attempts exceeds the lockout threshold (timespan minutes).")] String LockoutDuration;
```
Comment (X-Guardian) [inline source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof:10]:
```suggestion
    [Write, Description("Specifies the length of time that an account is locked after the number of failed login attempts exceeds the lockout threshold. The lockout duration must be greater than or equal to the lockout observation time for a password policy. The value must be a string representation of a TimeSpan value.")] String LockoutDuration;
```

File: source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    [Write, Description("Specifies the maximum time interval between two unsuccessful login attempts before the number of unsuccessful login attempts is reset to 0 (timespan minutes).")] String LockoutObservationWindow;
```
Comment (X-Guardian) [inline source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof:11]:
```suggestion
    [Write, Description("Specifies the maximum time interval between two unsuccessful login attempts before the number of unsuccessful login attempts is reset to 0. The lockout observation window must be smaller than or equal to the lockout duration for a password policy. The value must be a string representation of a TimeSpan value.")] String LockoutObservationWindow;
```

File: source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    [Write, Description("Specifies the number of unsuccessful login attempts that are permitted before an account is locked out.")] UInt32 LockoutThreshold;
+    [Write, Description("Specifies the minimum length of time before you can change a password (timespan days).")] String MinPasswordAge;
```
Comment (X-Guardian) [inline source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof:13]:
```suggestion
    [Write, Description("Specifies the minimum length of time before you can change a password. The value must be a string representation of a TimeSpan value.")] String MinPasswordAge;
```

File: source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    [Write, Description("Specifies the maximum length of time that you can have the same password (timespan days).")] String MaxPasswordAge;
```
Comment (X-Guardian) [inline source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof:14]:
```suggestion
    [Write, Description("Specifies the maximum length of time that you can have the same password. The value must be a string representation of a TimeSpan value.")] String MaxPasswordAge;
```

File: source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1 [hunk 5]
```diff
@@ -0,0 +1,802 @@
...
+        Specifies the ADPrincipal names the policy is to be applied to, overwrites all existing.
+
+    .PARAMETER Ensure
+        Specifies whether the fine grained password policy should be present or absent. Default value is 'Present'.
+
+    .PARAMETER ComplexityEnabled
+        Specifies whether password complexity is enabled for the password policy.
+
+    .PARAMETER LockoutDuration
+        Specifies the length of time that an account is locked after the number of failed login attempts exceeds the
```
Comment (X-Guardian) [inline source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1:167]: Please adjust the `LockoutDuration`, `LockOutObservationWindow`, `MinPasswordAge` and `MaxPasswordAge` descriptions to the new ones specified in the schema in both the `Test` and `Set` functions.

File: source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/en-US/about_ADFineGrainedPasswordPolicy.help.txt [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+.PARAMETER Ensure
+    Write - String
+    Allowed values: Present, Absent
+    Specifies whether the fine-grained password policy should be present or absent. Default value is 'Present'.
+
+.PARAMETER ComplexityEnabled
+    Write - Boolean
+    Specifies whether password complexity is enabled for the password policy.
+
+.PARAMETER LockoutDuration
```
Comment (X-Guardian) [inline source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/en-US/about_ADFineGrainedPasswordPolicy.help.txt:36]: Please adjust the `LockoutDuration`, `LockOutObservationWindow`, `MinPasswordAge` and `MaxPasswordAge` descriptions to the new ones specified in the schema.

**Comment by X-Guardian** [comment]
I just noticed your Unit and Integration tests are in a directory `Tests` with a capital 'T', where they should be in a directory `tests` with a lower case 't'. Windows will merge this into one, so you won't see it on your local PC, but you can see they are different if you look here. https://github.com/fl0p3k/ActiveDirectoryDsc/tree/newADFineGrainedPasswordPolicy. To fix them to lower case, I think you need to delete the files, commit the change, then create them again in the `test` directory with a lowercase 't' then commit them again.

**Comment by johlju** [comment]
It is possible to fix by using `git mv`. Use `git ls-files` to find the relative paths that should be changed then use for example:

```
git mv Tests/Unit/test.ps1 tests/Unit/test2.ps1
git mv tests/Unit/test2.ps1 tests/Unit/test.ps1
```

Need to rename the file to a temporary filename then back. After all files have been “moved” then rename the local folder in Windows file system.

**Comment by X-Guardian** [comment]
Well done @fl0p3k, we finally got there. Thanks for your patience and perseverance! We have ended up with a good quality resource.

**Comment by johlju** [comment]
@fl0p3k really impressive work on this one! Thank you! 😃

**Comment by fl0p3k** [comment]
Thanks for the review feedback and patience.  We made a lot of changes and learned a lot. I am grateful to have been a part of the process to get this done.

**Pull request merged**